### DISCUSSION TITLE
Certificate/Key Rotation

### DISCUSSION THREAD

**Discussion by MikeDombo** [emqx/emqx#7275] [Q&A] [answered]
Hi There!

If I'm running EMQ X and I need to change the certificate or key due to expiration or compromise would EMQ X dynamically reload them from the file system? If EMQ X is running with some clients connected and I then replace the values inside the certificate or key files, will _newly_ connecting clients receive the new certificate or old certificate?

**Comment by thalesmg** [comment, marked as answer]
Hi @MikeDombo !

If the certificates are changed, the Erlang VM will eventually reload those by itself.  That can take a few minutes to happen.  After that, new clients will see the new certificate.

Since 4.3.13, there is a [command](https://github.com/emqx/emqx/blob/main-v4.4/CHANGES-4.3.md#enhancements) that can make that happen at once: `emqx_ctl pem_cache clean`.

**Reply by MikeDombo** [reply]
Thanks @thalesmg that's fantastic! Is there somewhere in the code where I can see that logic and what defines when the VM would reload it?

Additionally, what happens to currently connected clients? Would they be disconnected during the reload?

**Reply by thalesmg** [reply]
The code for the `emqx_ctl` command can be found [here](https://github.com/emqx/emqx/blob/d58bfed751496199a2080d9ccb0833f7a7f75934/apps/emqx_management/src/emqx_mgmt.erl#L287-L294).

But the detail about when exactly the Erlang VM decides to reload the certificates pertains to OTP.  I don't know all details, but a possible good start may be the [`ssl_pem_cache`](https://github.com/erlang/otp/blob/21e8c995491aabe9d116d81dfe9af4a0773cbef2/lib/ssl/src/ssl_pem_cache.erl#L136) module.

The connections already established should not be affected by the reload.

**Reply by MikeDombo** [reply]
Is it possible to set application environment options from the EMQ X configuration or via environment variables? How often OTP checks for PEM changes is every 2 minutes by default, but it can be changed by setting the `ssl_pem_cache_clean` option of the `ssl` application. So is there a way to set this without modifying EMQ X?

https://github.com/erlang/otp/blob/21e8c995491aabe9d116d81dfe9af4a0773cbef2/lib/ssl/src/ssl_pem_cache.erl#L243-L244

**Reply by thalesmg** [reply]
That `application:get_env` is not reading from the environment variable, but rather from the [application config](https://www.erlang.org/doc/man/application.html#get_env-2).

Since the `ssl` application starts quite early in the boot process, that config would need to be set before that.  Maybe one could do it by adding it to `etc/vm.args` in the EMQX release dir.

But beware that this `ssl_pem_cache` API is not documented by OTP.  So it might change between OTP releases and break if one depends on this particular behavior.

**Reply by thalesmg** [reply]
By the way, the line that should be added is something like:

```
-ssl ssl_pem_cache_clean 999999
```

**Reply by MikeDombo** [reply]
Got it thank you very much again. We'll give it a go
